File: C++/LinkedList/LinkedList.cpp
#include <iostream>
#include <string>
using namespace std;

// A node in a linked list********************
class Node
{
public:
  Node(int value)
  {
    this->value = value;
    next = nullptr;
  }
  int value;
  Node *next;
};
// The linked list******************************
class LinkedList
{
public:
  LinkedList(int value)
  {
    Node *newNode = new Node(value);
    head = newNode;
    tail = newNode;
    length++;
  }
  void printList()
  {
    if (length == 0)
    {
      cout << "Empty Linked List";
      return;
    }
    Node *temp = head;
    while (temp)
    {
      cout << temp->value << endl;
      temp = temp->next;
    }
  }
  void getHead()
  {
    cout << head->value << endl;
  }
  void getTail()
  {
    cout << tail->value << endl;
  }
  void getLength()
  {
    cout << "Length: " << length << endl;
  }

  void append(int value)
  {
    Node *newNode = new Node(value);
    if (length == 0)
    {
      head = newNode;
      tail = newNode;
    }
    else
    {
      tail->next = newNode;
      tail = newNode;
    }
    length++;
  }

  void deleteLast()
  {
    if (length == 0)
      return;

    if (length == 1)
    {
      delete head;
      head = nullptr;
      tail = nullptr;
    }
    else
    {
      Node *prev = head;
      Node *next = head;
      while (next->next)
      {
        prev = next;
        next = next->next;
      }
      tail = prev;
      tail->next = nullptr;
      delete (next);
    }
    length--;
  }

  void prepend(int value)
  {
    Node *newNode = new Node(value);
    if (length == 0)
    {
      newNode->next = nullptr;
      head = newNode;
      tail = newNode;
    }
    else
    {
      newNode->next = head;
      head = newNode;
    }
    length++;
  }

  void deleteFirst()
  {
    if (length == 0)
      return;
    Node *temp = head;
    if (length == 1)
    {
      head = nullptr;
      tail = nullptr;
    }
    else
    {
      head = head->next;
    }
    delete (temp);
    length--;
  }

  Node *get(int index)
  {
    if (!validIndex(index))
      return nullptr;
    Node *temp = head;
    int i = 0;
    while (i != index)
    {
      temp = temp->next;
      i++;
    }
    return temp;
  }

  bool set(int index, int value)
  {
    if (index < 0 || index > length - 1)
      return false;

    Node *temp = head;
    int i = 0;
    while (i != index)
    {
      temp = temp->next;
      i++;
    }
    temp->value = value;

    return true;
  }

  bool insert(int index, int value)
  {
    if (index < 0 || index > length)
    {
      cout << "Invalid Index" << endl;
      return false;
    }

    if (index == 0)
    {
      prepend(value);
      return true;
    }

    if (index == length)
    {
      append(value);
      return true;
    }

    Node *newNode = new Node(value);
    Node *prev = get(index - 1);
    newNode->next = prev->next;
    prev->next = newNode;
    length++;
    return true;
  }

  void deleteNode(int index)
  {
    if (!validIndex(index))
      return;

    if (index == 0)
      return deleteFirst();

    if (index == length - 1)
      return deleteLast();

    Node *prev = get(index - 1);
    Node *temp = prev->next;
    prev->next = temp->next;
    delete temp;

    length--;
  }

  void reverse()
  {
    if (length < 2)
      return;

    Node *temp = head;
    head = tail;
    tail = temp;

    Node *after = temp->next;
    Node *before = nullptr;

    for (int i = 0; i < length; i++)
    {
      after = temp->next;
      temp->next = before;
      before = temp;
      temp = after;
    }
  }

  ~LinkedList()
  {
    Node *temp = head;
    while (head)
    {
      head = head->next;
      delete temp;
      temp = head;
    }
  }

private:
  Node *head;
  Node *tail;
  int length;
  bool validIndex(int index)
  {
    if (index < 0 || index > length - 1)
    {
      cout << "Error! Index out of bounds" << endl;
      return false;
    }
    return true;
  }

  Node *nodeAtIndex(int index)
  {
    Node *temp = head;
    int i = 0;
    while (i != index)
    {
      temp = temp->next;
      i++;
    }
    return temp;
  }
};

int main()
{
  LinkedList *myList = new LinkedList(200);
  // myList->deleteNode(0);
  // myList->deleteNode(0);
  // myList->deleteFirst();
  // myList->deleteLast();
  myList->append(300);
  myList->prepend(100);
  myList->append(400);
  myList->append(500);
  myList->append(600);
  // myList->set(0, 12);
  // myList->insert(0, 12);

  // myList->reverse();
  myList->printList();
  return 0;
}

File: Python/LinkedList/LinkedList.py
class Node:
  def __init__(self, data) -> None:
    self.data = data
    self.next = None
    
class LinkedList:
  def __init__(self, data=None):
    self.head = None;
    self.tail = None;
    self.length = 0;
    if data is not None:
      self.append(data)
      
  def printList(self):
    if self.length == 0:
      print("Empty List")
      return
    temp = self.head;
    for i in range(self.length):
      print(temp.data)
      temp = temp.next
      
  def get(self, index)->Node:
    if index < 0 or index > self.length-1:
      raise ValueError("Index out of bounds")
    temp = self.head
    for i in range(index):
      temp = temp.next
    return temp
      
  def set(self, index, data):
    temp = self.get(index)
    temp.data = data
  
  def insert(self, index, data):
    if index < 0 or index > self.length:
      raise ValueError("Index out of bounds")
    if index == 0:
      self.prepend(data)
      return
    if index == self.length:
      self.append(data)
      return
    newNode = Node(data)
    prev = self.head
    temp = self.head
    for i in range(index):
      prev = temp
      temp = temp.next
    newNode.next = temp
    prev.next = newNode

    self.length += 1

  def append(self, data):
    node = Node(data)
    if self.length == 0:
      self.head = node
      self.tail = node
    self.tail.next = node;
    self.tail = node;
    self.length += 1
      
  def prepend(self,data):
    newNode = Node(data)
    if self.length == 0:
      self.head = newNode
      self.tail = newNode
    else:
      newNode.next = self.head
      self.head = newNode
    self.length += 1;
    
  def deleteLast(self):
    if self.length == 0: return
    if self.length ==  1:
      self.head = None
      self.tail = None
    else:
      temp = self.head
      while(temp.next.next):
        temp = temp.next
      node_to_delete = temp.next
      temp.next = None
      self.tail = temp
      del node_to_delete
    
    self.length -= 1;
    
  def deleteFirst(self):
    if self.length == 0:
      return 
    if self.length == 1:
      self.head = None
      self.tail = None
    else:
      temp = self.head;
      self.head = self.head.next;
      del temp
      
    self.length -= 1
      
  def deleteNode(self, index):
    if index < 0 or index > self.length -1: return print("Index out of bounds")
    if index == 0: return self.deleteFirst()
    if index == self.length - 1: return self.deleteLast()
    
    temp = self.head
    for i in range(index-1):
      temp = temp.next
    node_to_delete = temp.next;
    temp.next = temp.next.next
    del node_to_delete
    self.length -= 1
    
  def reverse(self):
    if self.length < 2: return print("Cannot reverse link less than 2 nodes")
    
    before = None
    current = self.head
    after = current.next
    
    self.head = self.tail
    self.tail = current
    
    for i in range(self.length):
      after = current.next
      current.next = before
      before = current
      current = after
      
    

#Code test**************************************
myLinkedList = LinkedList()
myLinkedList.append(20)
myLinkedList.append(30)
myLinkedList.prepend(5)
myLinkedList.set(1,9)
myLinkedList.insert(0,10000)
myLinkedList.printList()
myLinkedList.deleteNode(3)
print("-------------")
myLinkedList.reverse()
myLinkedList.printList()